lib/logstash/plugin_mixins/config_deprecation_support.rb

```diff
@@ -0,0 +1,74 @@
+# encoding: utf-8
+
+require 'logstash/namespace'
+require 'logstash/plugin'
+
+module LogStash
+  module PluginMixins
+    ##
+    # This `ConfigDeprecationSupport` can be included in any `LogStash::Plugin`,
+    # and will provide utilities methods that can be used by the plugins to
+    # extract and deprecate configs.
+    module ConfigDeprecationSupport
+      ##
+      # @api internal (use: `LogStash::Plugin::include`)
+      # @param base [Class]: a class that inherits `LogStash::Plugin`, typically one
+      #                      descending from one of the four plugin base classes
+      #                      (e.g., `LogStash::Inputs::Base`)
+      # @return [void]
+      def self.included(base)
+        fail(ArgumentError, "`#{base}` must inherit LogStash::Plugin") unless base < LogStash::Plugin
+      end
+
+      ##
+      # Unambiguously extracts the value of a param that may be provided with one or more deprecated params
+      #
+      # The `value_transformer` block is used when one or more deprecated params are explicitly provided,
+      # to transform their effective values into a single suitable value for use as-if it had been
+      # provided by the preferred param.
+      # It is required except in the case of simple param renames.
+      #
+      # @param preferred_param [String]: the preferred param name
+      # @param deprecated_params [String...]: the deprecated param names
+      # @yield values_from_deprecated_params [Object...]: the ordered, validated-and-transformed values of _all_
+      #                                                   deprecated params, including default values.
+      # @yieldreturn [Object]: a single value to use as-if it had been provided by the preferred param
+      #
+      # @raise `LogStash::ConfigurationError` if both preferred and deprecated params are explicitly provided
+      # @return [Object]: the value of the preferred param or an equivalent derived from provided deprecated params
+      #
+      # @note Relies on upstream deprecation warnings, and does not emit its own
+      def config_with_deprecated_target!(preferred_param, *deprecated_params, &value_transformer)
```

Comment:
"config with deprecated target" doesn't clearly convey to me what this method _does_, partially because "deprecated target" implies to me that the place where this config is going _to_ is deprecated.

Focusing on how it will be used, I think we may be best off defining a method `normalize_config` that takes a single parameter (the name of the config to produce a normalized value for) and yields a configurator block that can be used to configure the normalization. That configurator could have methods like `with_deprecated_mapping` or `with_deprecated_alias` that map or alias deprecated config counterparts:

~~~ ruby
  @ssl_verification_mode = normalize_config(:ssl_verification_mode) do |normalize|
    normalize.with_deprecated_mapping(:ssl_verify_mode) do |ssl_verify_mode|
      case ssl_verify_mode
      when "none" then "none"
      when "peer" then "certificate"
      when "force_peer" then "full"
      else fail(LogStash::ConfigurationError, "Unsupported value #{ssl_verify_mode} for deprecated option `ssl_verify_mode`")
    end
  end

  @ssl_cipher_suites = normalize_config(:ssl_cipher_suites) do |normalize|
    normalize.with_deprecated_alias(:cipher_suites)
  end

  @ssl_supported_protocols = normalize_config(:ssl_supported_protocols) do |normalize|
    normalize.with_deprecated_mapping(:tls_min_version, :tls_max_version) do |tls_min, tls_max|
      TLS.get_supported(tls_min..tls_max).map(&:name)
    end
  end
~~~

I would be glad to tackle the implementation, since it does require a bit of advanced meta-programming, but at a high-level it would be something like:

~~~ ruby
module NormalizeConfigSupport
  def normalize_config(canonical_config, &configurator)
    ConfigNormalizer.new(self, canonical_config, &configurator).value
  end

  class ConfigNormalizer
    def initialize(plugin, canonical_config, &configurator)
      @plugin = plugin
      @canonical_config = canonical_config.to_s

      @deprecated_mappings = nil

      yield(self)
    end

    def with_deprecated_mapping(*deprecated_params, &value_transformer)
      # TODO: ensure params exist and actually are deprecated (using `@plugin.class.get_config`)
      # TODO: ensure value transformer arity matches
      # TODO: ensure deprecated_mappings isn't already set
      @deprecated_mappings = [deprecated_params.map(&:to_s), value_transformer]
    end

    def with_deprecated_alias(deprecated_alias)
       # TODO: ensure alias exists and actually is deprecated (using `@plugin.class.get_config`)
       with_deprecated_mapping(deprecated_alias) { |v| v }
    end

    def value
      # TODO: ensure canonical and deprecated not both supplied (using `@plugin.original_params`)
      # TODO: use @deprecated_mappings to extract effective value
    end
  end
  private_constant :ConfigNormalizer
end
~~~

The benefit I see here is that we present a very limited and succinct new API `normalize_config` that clearly conveys our intent to normalize a config parameter and provides us with what we currently need (the ability to wholly-alias a deprecated config, the ability to map one or more deprecated configs to the canonical value), while also giving us room for more if and when we discover it, all without polluting the method namespace of the plugins that this support adapter is mixed into. For example, if at some point in the future we needed to deprecate individual _values_ for a config setting, we could introduce a method on our `ConfigNormalizer`, without forcing it to be added to all of the plugins that use this support adapter.

Reply:
Thank you very much for your review @yaauie! that looks great and makes total sense as we're likely to add more methods to this helper in the future. I'll apply your suggestions/implementation.

Reply:
> ~~Considering the suggested namings, I think we should also rename this project/repository to something like `logstash-mixin-config_normalize_support`. WDYT?~~

Even though you crossed it out, I think it does more-accurately describe the intent of this support adapter. My "NormalizeConfigSupport" was just pseudo-code, but any of the following would describe the intent in a way that aligns with our other support adapters:
 - `NormalizeConfigSupport`
   - If it provides declarative `normalize_config` there is a nice 1:1 mapping between module name and provided method
 - `ConfigNormalizerSupport`
   - If it provides declarative `normalize_config`, the mapping isn't quite 1:1 but still a very tight association between the name of the module that is mixed in and the API it provides
   - If it provides the 1:1 mapped `config_normalize`, the method name does read left-to-right in a way that builds context, but isn't quite as declarative in my mind

Reply:
My first thought was renaming this repository & gem to  `logstash-mixin-config_normalize_support`, but after a second thought, I realize that the `-config_normalize_support` wasn't very intuitive for me without the `deprecation` context. That's why I ~~crossed~~ it out. thoughts?

I did like your pseudo-code namings actually! I've pushed the code already with `NormalizeConfigSupport` (before reading this comment).  I still need to double-check everything and test the new changes with a real plugin. But feel free to have a second look :)

Reply:
Yes, it sounds good to me! I'll rename this repository/project/etc to `logstash-mixin-normalize_config_support` then.